### PR TITLE
[MPM] consistent mass matrix

### PR DESCRIPTION
**Pull request by peterjwilson**
Adding consistent mass matrix option. The default is still lumped. The consistent mass matrix dissipates less energy than the lumped, and, for the testing done, did not appear to add noticeable computation time (since K already has off-trace terms in an implicit solve):

image

### CODE REVIEW

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp [hunk 1]
```diff
@@ -1019,35 +1019,44 @@ void UpdatedLagrangianUP::GetDofList( DofsVectorType& rElementalDofList, Process
...
+        ? rCurrentProcessInfo.GetValue(USE_CONSISTENT_MASS_MATRIX)
+        : false;
 
-    if ( rMassMatrix.size1() != matrix_size )
-        rMassMatrix.resize( matrix_size, matrix_size, false );
+    const SizeType dimension = GetGeometry().WorkingSpaceDimension();
+    const SizeType number_of_nodes = GetGeometry().PointsNumber();
+    const SizeType matrix_size = dimension * number_of_nodes;
 
+    if (rMassMatrix.size1() != matrix_size)
```
Comment (tteschemacher) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp]: don't you need to check also size2?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp [hunk 2]
```diff
@@ -1276,36 +1276,44 @@ void UpdatedLagrangian::CalculateDampingMatrix( MatrixType& rDampingMatrix, Proc
...
-    unsigned int matrix_size = dimension * number_of_nodes;
+    const bool is_consistent_mass_matrix = (rCurrentProcessInfo.Has(USE_CONSISTENT_MASS_MATRIX))
+        ? rCurrentProcessInfo.GetValue(USE_CONSISTENT_MASS_MATRIX)
+        : false;
+
+    const SizeType dimension = GetGeometry().WorkingSpaceDimension();
+    const SizeType number_of_nodes = GetGeometry().PointsNumber();
+    const SizeType matrix_size = dimension * number_of_nodes;
 
     if ( rMassMatrix.size1() != matrix_size )
```
Comment (tteschemacher) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp]: size2?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_quadrilateral.cpp [hunk 3]
```diff
@@ -1277,33 +1277,44 @@ void UpdatedLagrangianQuadrilateral::CalculateDampingMatrix( MatrixType& rDampin
...
+        ? rCurrentProcessInfo.GetValue(USE_CONSISTENT_MASS_MATRIX)
+        : false;
 
-    if ( rMassMatrix.size1() != matrix_size )
-        rMassMatrix.resize( matrix_size, matrix_size, false );
+    const SizeType dimension = GetGeometry().WorkingSpaceDimension();
+    const SizeType number_of_nodes = GetGeometry().PointsNumber();
+    const SizeType matrix_size = dimension * number_of_nodes;
 
+    if (rMassMatrix.size1() != matrix_size)
```
Comment (tteschemacher) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_quadrilateral.cpp]: same

**Review by tteschemacher — COMMENTED** [review]
I personally like the way that you did it!

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp (hunk 1 above)
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp]: yep, done

**Review by tteschemacher — APPROVED** [review]
Thank you @PeterJosephWilson for the changes! It looks good to me and provides a nice new feature!

**Review by bodhinandach — APPROVED** [review]
I am good too! Excellent @peterjwilson. Many thanks

**Comment by antonialarese** [comment]
Thanks Peter! Nice work

**Comment by philbucher** [comment]
I didn't see this but just to say that this is inconsistent with what StructuralMechanics is doing, there `COMPUTE_LUMPED_MASS_MATRIX` from the core is used (#3802)

=> This hasn't been updated yet in all elements, but eventually it will be updated
=> I recommend doing the same

**Comment by peterjwilson** [comment]
> I didn't see this but just to say that this is inconsistent with what StructuralMechanics is doing, there `COMPUTE_LUMPED_MASS_MATRIX` from the core is used (#3802)
> 
> => This hasn't been updated yet in all elements, but eventually it will be updated
> => I recommend doing the same

Thanks for the comment. I must have picked the wrong random structural element to check first :). New PR is open for the change.

**Pull request merged**